redfish-core/lib/oem/ibm/pcie_topology_refresh.hpp

```diff
@@ -0,0 +1,154 @@
+        (*countPtr) = 0;
+        return;
+    }
+    //  This method can block incoming requests max for 8 seconds. So, each
+    //  call to this method adds a 1-second block, and the maximum call allow is
+    //  8 times which makes it a total of ~8 seconds
+    if ((*countPtr) >= 8)
+    {
+        messages::internalError(asyncResp->res);
+        pcieTopologyRefreshTimer = nullptr;
+        (*countPtr) = 0;
+        return;
+    }
+    ++(*countPtr);
+    sdbusplus::asio::getProperty<bool>(
+        *crow::connections::systemBus, "xyz.openbmc_project.PLDM",
+        "/xyz/openbmc_project/pldm", "com.ibm.PLDM.PCIeTopology",
+        "PCIeTopologyRefresh",
+        [asyncResp, timer, countPtr](const boost::system::error_code& ec1,
+                                     const bool pcieRefreshValue) {
+        if (ec1)
+        {
+            BMCWEB_LOG_ERROR("DBUS response error {}", ec1.value());
+            messages::internalError(asyncResp->res);
+            pcieTopologyRefreshTimer = nullptr;
+            (*countPtr) = 0;
+            return;
+        }
+        // After PCIe Topology Refresh, it sets the pcieRefreshValuePtr
+        // value to false. if a value is not false, extend the time, and if
+        // it is false, delete the timer and reset the counter
+        if (pcieRefreshValue)
+        {
+            BMCWEB_LOG_ERROR("pcieRefreshValuePtr time extended");
+            timer->expires_at(timer->expiry() +
+                              boost::asio::chrono::seconds(1));
+            timer->async_wait([timer, asyncResp,
+                               countPtr](const boost::system::error_code ec2) {
+                pcieTopologyRefreshWatchdog(ec2, timer, asyncResp, countPtr);
+            });
+        }
+        else
+        {
+            BMCWEB_LOG_ERROR("pcieRefreshValuePtr value refreshed");
+            pcieTopologyRefreshTimer = nullptr;
+            (*countPtr) = 0;
+            return;
+        }
+    });
+};
+
+/**
+ * @brief Sets PCIe Topology Refresh state.
+ *
+ * @param[in] asyncResp   Shared pointer for generating response message.
+ * @param[in] state   PCIe Topology Refresh state from request.
+ *
+ * @return None.
+ */
+inline void
+    setPCIeTopologyRefresh(const crow::Request& req,
+                           const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+                           const bool state)
+{
+    BMCWEB_LOG_DEBUG("Set PCIe Topology Refresh status.");
+    sdbusplus::asio::setProperty(
+        *crow::connections::systemBus, "xyz.openbmc_project.PLDM",
+        "/xyz/openbmc_project/pldm", "com.ibm.PLDM.PCIeTopology",
+        "PCIeTopologyRefresh", state,
+        [&req, asyncResp](const boost::system::error_code& ec) {
+        if (ec)
+        {
+            BMCWEB_LOG_ERROR("PCIe Topology Refresh failed.{}", ec.value());
+            messages::internalError(asyncResp->res);
+            return;
+        }
+        countPCIeTopologyRefresh = 0;
+        pcieTopologyRefreshTimer =
```

Comment:
What is making sure there isn't already a pcieTopologyRefreshTimer?

Reply:
This is to reset/restart the topology refresh timer when redfish `PCIeTopologyRefresh` is issued.
If `pcieTopologyRefreshTimer` was already started/used, this line will recreate a fresh new one (by removing the old one - i.e. by `std::unique_ptr` semantics).